packages/ember-data/lib/system/snapshot.js

```diff
@@ -0,0 +1,299 @@
+
+    `hasMany` takes an optional hash of options as a second parameter,
+    currently supported options are:
+
+   - `ids`: set to `true` if you only want the IDs of the related items to be
+      returned.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { title: 'Hello World', comments: [2, 3] });
+    var snapshot = post._snapshot();
+
+    snapshot.hasMany('comments'); // => [DS.Snapshot, DS.Snapshot]
+    snapshot.hasMany('comments', { ids: true }); // => [2, 3]
+    ```
+
+    Note: Relationships are loaded lazily and cached upon first access.
+
+    @method hasMany
+    @param {string} keyName
+    @param {Object} [options]
+    @return {Array} An array of snapshots of a hasMany relationship
+  */
+  hasMany: function(keyName, options) {
+    var ids = options && options.ids;
+    var snapshots = [];
+    var relationship, members, hasMany;
+
+    if (!(keyName in this._hasManyRelationships)) {
+      relationship = this.record._relationships[keyName];
+
+      if (!(relationship && relationship.relationshipMeta.kind === 'hasMany')) {
+        throw new Ember.Error("Model '" + Ember.inspect(this.record) + "' has no hasMany relationship named '" + keyName + "' defined.");
+      }
+
+      members = get(relationship, 'members');
+      members.forEach(function(member) {
+        snapshots.push(member._snapshot());
+      });
+
+      this._hasManyRelationships[keyName] = snapshots;
+    }
+
+    hasMany = this._hasManyRelationships[keyName];
+
+    return ids ? Ember.A(hasMany).mapBy('id') : hasMany;
+  },
+
+  /**
+    Iterates through all the attributes of the model, calling the passed
+    function on each attribute.
+
+    Example
+
+    ```javascript
+    snapshot.eachAttribute(function(name, meta) {
+      // ...
+    });
+    ```
+
+    @method eachAttribute
+    @param {Function} callback the callback to execute
+    @param {Object} [binding] the value to which the callback's `this` should be bound
+  */
+  eachAttribute: function(callback, binding) {
+    this.record.eachAttribute(callback, binding);
+  },
+
+  /**
+    Iterates through all the relationships of the model, calling the passed
+    function on each relationship.
+
+    Example
+
+    ```javascript
+    snapshot.eachRelationship(function(name, relationship) {
+      // ...
+    });
+    ```
+
+    @method eachRelationship
+    @param {Function} callback the callback to execute
+    @param {Object} [binding] the value to which the callback's `this` should be bound
+  */
+  eachRelationship: function(callback, binding) {
+    this.record.eachRelationship(callback, binding);
+  },
+
+  /**
+    @method get
+    @param {string} keyName
+    @return {Object} The property value
+    @deprecated Use [attr](#method_attr), [belongsTo](#method_belongsTo) or [hasMany](#method_hasMany) instead
+  */
+  get: function(keyName) {
+    Ember.deprecate('Using DS.Snapshot.get() is deprecated. Use .attr(), .belongsTo() or .hasMany() instead.');
```

Comment:
This seems like a completely new API, why would we come out of the gates with deprecations?

Pardon me if this has already been discussed...